numpy/core/src/multiarray/arrayobject.c

```diff
@@ -661,36 +534,42 @@ static PyObject *
 array_repr(PyArrayObject *self)
 {
     PyObject *s, *arglist;
+    PyObject *reprfunc = NULL;
 
-    if (PyArray_ReprFunction == NULL) {
-        s = array_repr_builtin(self, 1);
+    npy_cache_import("numpy.core._internal", "_pyarray_repr_function", &reprfunc);
```

Comment:
The goal is to store the str and repr functions within python so that the lifecycle of the functions is tied to the lifecycle of the interpreter(or subinterpreter). 
1) npy_cache_import only returns NULL if there was an error loading the module or an error looking up the value within the module. Both of these conditions set a python error which will get propagated to the caller.
2) Since reprfunc is not a static variable, the cache import will import the module and lookup the variable every time this is called.